node/src/actors/session/handlers.rs

```diff
@@ -386,14 +412,37 @@ fn peer_discovery_peers(peers: &[Address]) {
 }
 
 /// Function called when Block message is received
-fn inventory_process_block(_session: &mut Session, _ctx: &mut Context<Session>, block: Block) {
+fn inventory_process_block(session: &mut Session, _ctx: &mut Context<Session>, block: Block) {
     // Get ChainManager address
     let chain_manager_addr = System::current().registry().get::<ChainManager>();
 
-    // Send a message to the ChainManager to try to add a new block
-    chain_manager_addr.do_send(AddBlocks {
-        blocks: vec![block],
-    });
+    let block_epoch = block.block_header.beacon.checkpoint;
+    let block_hash = block.hash();
+
+    if Some(block_epoch) == session.current_epoch {
+        // Send a message to the ChainManager to try to add a new candidate
+        chain_manager_addr.do_send(AddCandidates {
+            blocks: vec![block],
+        });
+    // TODO: Add timeout
+    } else if session.requested_blocks.len() == session.requested_block_hashes.len() {
+        let mut blocks_vector = vec![];
+        for hash in session.requested_block_hashes.clone() {
+            if let Some(block) = session.requested_blocks.remove(&hash) {
+                blocks_vector.push(block);
+            } else {
+                break;
+            }
+        }
+
+        // Send a message to the ChainManager to try to add a new block
+        chain_manager_addr.do_send(AddBlocks {
+            blocks: blocks_vector,
+        });
+    } else {
+        // Add block to requested_blocks
+        session.requested_blocks.insert(block_hash, block);
```

Comment:
Here we are storing all the blocks, regardless if we requested them or not. This could be a good way to cache blocks, but since this map is unique to each session, we probably won't need the unrequested blocks.